### ISSUE TITLE
[Question] iterative refinement prone to overfit

### ISSUE THREAD

**Issue by NatureGeorge** [RalphHan/CASR#1]
Hi. Your idea of modeling complex sequences via so-called self-boost refinement is very interesting.

I tried to implement this approach to neural machine translation. That is, train an autoregressive model $M^{0}$ to translate a source sequence to a target sequence in the teacher-forcing manner. After training enough, generate the predicted version of the target sequences. Then, take the source sequence and predicted target sequence as input to the second autoregressive model $M^{1}$ and still train it in the teacher-forcing manner. And so on.

However, I found that, since this is not an end-to-end procedure, the validation loss goes higher as the `Castep` increases. The expectation that `the later model corrects the wrong part of the previous prediction according to other parts it depends on` fails. This result is not a surprise to me because the model can easily find a shortcut which is just simply copying the inputs and this shortcut conveniently maintains a good likelihood loss. In other words, in my implementation, only the first model $M^{0}$ learns something, and the subsequent model $\lbrace M^{t}\rbrace_{t>0}$ just overfits the input and is too lazy to try to correct the wrong predictions.

So I am wondering how exactly CASR model can achieve `refinement`.

**Comment by RalphHan** [comment]
Hi, thanks for your interest! 

We found the increment becomes marginal when Castep increases, on all tasks, especially NLG tasks. Two possible reasons:
1. NLG tasks like translation are not complex enough compared with code generation. 
2. We don't increase temperature (noise) in early Casteps, making the model lazy and lack exploration. From now perspective, we think CASR is similar with diffusion model. In the denoise process of diffusion, the variance becomes smaller and smaller, while in the generation process of CASR, the same beam search sampling method is used. And in diffusion, all steps use the same model (which treat timestep embedding as input), in CASR, different steps use different models (not input time embedding). 

I hope this may inspire you,
Thank you again for your attention!

**Comment by NatureGeorge** [comment]
Appreciate your quick and kind reply. Start from a high temperature (for sampling) and gradually decrease it seems to be a promising direction. I would try it.

**Comment by NatureGeorge** [comment]
Just for a reminder, I found that this is basically the Order-Agnostic Autoregressive Diffusion Model.
